cpp/src/arrow/compute/kernel.h

```diff
@@ -43,25 +43,64 @@ class ARROW_EXPORT OpKernel {
 
 /// \brief Placeholder for Scalar values until we implement these
 struct ARROW_EXPORT Scalar {
-  ~Scalar() {}
-
-  ARROW_DISALLOW_COPY_AND_ASSIGN(Scalar);
+  util::variant<bool, uint8_t, int8_t, uint16_t, int16_t, uint32_t, int32_t, uint64_t, int64_t, float, double> value;
+
+  Scalar(bool value) : value(value) {}
+  Scalar(uint8_t value) : value(value) {}
+  Scalar(int8_t value) : value(value) {}
+  Scalar(uint16_t value) : value(value) {}
+  Scalar(int16_t value) : value(value) {}
+  Scalar(uint32_t value) : value(value) {}
+  Scalar(int32_t value) : value(value) {}
+  Scalar(uint64_t value) : value(value) {}
+  Scalar(int64_t value) : value(value) {}
+  Scalar(float value) : value(value) {}
+  Scalar(double value) : value(value) {}
+
+  Type::type kind() const {
+    switch (this->value.which()) {
+      case 0:
```

Comment:
Couldn't you just `static_cast<Type::type>(this->value.which())`?

Reply:
We'd have to make sure the mapping is exact, which is almost equivalent to expliciting this list.